### PR TITLE
[libc++] The rest of P2278R4 (cdata, as_const_view, span changes)

### PR DESCRIPTION
**Pull request by strega-nil**
This PR depends on, and is committed on top of, [LLVM-PR-99915][].

This PR completes the rest of [P2278R4][] that was not done in the original PR:

Actual changes (excluding those from LLVM-PR-99915): [diff 99915][].
- `cdata`
- Changes to `span` (includes some interesting choices since `const_iterator` cannot support `volatile` element types)
- `as_const_view` and `views::as_const`
- The `__cpp_lib_ranges_as_const` feature macro

[LLVM-PR-99915]: https://github.com/llvm/llvm-project/pull/99915
[P2278R4]: https://wg21.link/p2278r4
[diff 99915]: https://github.com/llvm/llvm-project/pull/102301/files/54f6a6e50f6bb8fbaffecf99ea59d394beb37658..61ac4bb0e76643036150f26c6f95e7492a4da1e6

### CODE REVIEW

File: libcxx/include/__ranges/as_const_view.h [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  using type = _XType;
+};
+
+struct __fn : __range_adaptor_closure<__fn> {
+  // implementation strategy taken from Microsoft's STL
+  enum class __strategy {
+    __already_const,
+    __empty_view,
+    __span,
+    __ref_view,
```
Comment (frederick-vs-ja) [inline libcxx/include/__ranges/as_const_view.h:135]: [LWG3811](https://cplusplus.github.io/LWG/issue3811) and [LWG3850](https://cplusplus.github.io/LWG/issue3850) are completed here and thus the entries should be updated:
https://github.com/llvm/llvm-project/blob/445023f1739d3764f1dae6e6b7b0602328198a1e/libcxx/docs/Status/Cxx23Issues.csv#L257
https://github.com/llvm/llvm-project/blob/445023f1739d3764f1dae6e6b7b0602328198a1e/libcxx/docs/Status/Cxx23Issues.csv#L272

File: libcxx/include/__ranges/const_access.h [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+} // namespace ranges
+
+// [range.prim.cdata]
+
+namespace ranges {
+namespace __cdata {
+struct __fn {
+#  if _LIBCPP_STD_VER >= 23
+  template <class _Tp>
+  _LIBCPP_HIDE_FROM_ABI constexpr static auto __as_const_pointer(const _Tp* __ptr) noexcept {
```
Comment (frederick-vs-ja) [inline libcxx/include/__ranges/const_access.h:211]: Ditto for [LWG3948](https://cplusplus.github.io/LWG/issue3948) (assuming the #99915 will be merged first):
https://github.com/llvm/llvm-project/blob/445023f1739d3764f1dae6e6b7b0602328198a1e/libcxx/docs/Status/Cxx2cIssues.csv#L30

File: libcxx/include/__ranges/view_interface.h [hunk 3]
```diff
@@ -72,6 +73,27 @@ class view_interface {
     }
   }
 
+  constexpr auto cbegin()
+    requires input_range<_Derived>
```
Comment (frederick-vs-ja) [inline libcxx/include/__ranges/view_interface.h:77]: Ditto for [LWG3766](https://cplusplus.github.io/LWG/issue3766):
https://github.com/llvm/llvm-project/blob/445023f1739d3764f1dae6e6b7b0602328198a1e/libcxx/docs/Status/Cxx23Issues.csv#L223

**Review by frederick-vs-ja — CHANGES_REQUESTED** [review]

**Comment by strega-nil** [comment]
Closing as this will be merged into the original PR.

**Pull request closed without merging**